File: packages/contract/src/parse-guards.ts
export function recordOf(value: unknown): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error('Contrato inválido');
  }
  return value as Record<string, unknown>;
}

export function textOf(value: unknown): string {
  if (typeof value !== 'string' || value.trim() === '') {
    throw new Error('Contrato inválido');
  }
  return value;
}

export function intOf(value: unknown): number {
  if (typeof value !== 'number' || !Number.isInteger(value) || value < 1) {
    throw new Error('Contrato inválido');
  }
  return value;
}

export function stringList(value: unknown): string[] {
  if (!Array.isArray(value) || value.some((item) => typeof item !== 'string')) {
    throw new Error('Contrato inválido');
  }
  return value;
}

File: apps/mobile/lib/contract/parse_guards.dart
const contractError = 'Contrato inválido';

Map<String, Object?> recordOf(Object? value) {
  if (value is! Map) throw const FormatException(contractError);
  return {for (final entry in value.entries) entry.key.toString(): entry.value as Object?};
}

String textOf(Object? value) {
  if (value is! String) throw const FormatException(contractError);
  if (value.trim().isEmpty) throw const FormatException(contractError);
  return value;
}

String stringItem(Object? value) {
  if (value is! String) throw const FormatException(contractError);
  return value;
}

int intOf(Object? value) {
  if (value is! int) throw const FormatException(contractError);
  if (value < 1) throw const FormatException(contractError);
  return value;
}

int countOf(Object? value) {
  if (value is! int) throw const FormatException(contractError);
  if (value < 0) throw const FormatException(contractError);
  return value;
}

List<String> stringList(Object? value) {
  if (value is! List) throw const FormatException(contractError);
  return [for (final item in value) stringItem(item as Object?)];
}
